Fix check in EventDataBatch.TryAdd, which could result in a batch exceeding its max size

illfygli opened the pull request:
It was possible to add a single `EventData` that exceeded the max size of the batch, due to the `this.eventDataList.Count > 0` condition, which I removed.

serkantkaraca commented:
The count check is there to allow addition of first message that is larger than 256 KB.

illfygli commented:
@serkantkaraca Should it not respect the `maxSizeInBytes` parameter, as [indicated by the documentation](https://docs.microsoft.com/en-us/dotnet/api/microsoft.azure.eventhubs.eventdatabatch.-ctor?view=azure-dotnet#parameters)?

serkantkaraca commented:
> @serkantkaraca Should it not respect the `maxSizeInBytes` parameter, as [indicated by the documentation](https://docs.microsoft.com/en-us/dotnet/api/microsoft.azure.eventhubs.eventdatabatch.-ctor?view=azure-dotnet#parameters)?

The default is 256 KB when created on a client that has no connection to the service yet. 
https://github.com/Azure/azure-sdk-for-net/blob/d2075bdf3e1679e5660742d7ca2735c0a7a59398/sdk/eventhub/Microsoft.Azure.EventHubs/src/Amqp/AmqpEventDataSender.cs#L27

illfygli commented:
But that's not the max batch size, that's determined [here](https://github.com/Azure/azure-sdk-for-net/blob/0b4d56eba9bd85d6efb0234582de6768364bfa76/sdk/eventhub/Microsoft.Azure.EventHubs/src/EventDataBatch.cs#L37).

If the `maxSizeInBytes` parameter for `EventDataBatch` does not specify the max size of the batch, what is it actually for?

I ran into this when fixing an issue where batches were sent, even though they exceeded the specified max size, which `TryAdd` should prevent, according to the docs.

illfygli commented:
With both Standard SKU and Basic SKU, you can currently exceed the max size, using `TryAdd`.

You can add a 2MB message, which exceeds the 1MB limit of Standard.

As far as I can tell, this will keep happening after the link has been established as well, since the size check is not performed at all for the first object.

serkantkaraca commented:
Can you fix it in a way so batchOptions.maxSize can override the default and skip the first message check?

illfygli commented:
> Can you fix it in a way so batchOptions.maxSize can override the default and skip the first message check?

I think that's [how it currently works](https://github.com/Azure/azure-sdk-for-net/blob/0b4d56eba9bd85d6efb0234582de6768364bfa76/sdk/eventhub/Microsoft.Azure.EventHubs/src/EventDataBatch.cs#L37), if I understand you correctly.

The `EventDataBatch` constructor sets `this.maxSize = Math.Min(userSuppliedMaxSize, HardCoded4MegMaxSize);`

illfygli commented:
Maybe `BatchOptions.MaxMessageSize` should have a default value too.

Looks like it defaults to 0, when left unspecified.

Due to the current check in `TryAdd`, each batch would only ever have one event in it, if the user does not specify a max size, since it would compare to 0 after one event has been added.

serkantkaraca commented:
BatchOptions.MaxMessageSize default is zero.

My proposal is to change EventDataBatch in a way that it can skip "first message" check if the size is set via BatchOptions. I think the cleanest solution may be to add an internal constructor to accept the user-defined size and skipFirstMessageCheck boolean.

illfygli commented:
Thanks, I will give it a try.

illfygli committed:
Fix `EventDataBatch.TryAdd` so that it respects `MaxMessageSize`

If no `MaxMessageSize` is explicitly given, it falls back to the old behavior,
where the first `EventData` is always accepted.

illfygli commented:
I implemented it as suggested, and added tests both for usage with an explicit max size, and without, when it falls back to the old behaviour of always accepting the first `EventData`.

illfygli commented:
Ping @serkantkaraca
